src/bin/pg_autoctl/monitor.c

```diff
@@ -1865,13 +1866,76 @@ prepare_connection_to_current_system_user(Monitor *source, Monitor *target)
 
 
 /*
- * monitor_listen_loop loops over a LISTEN command that is notified at every
+ * ensure_monitor_pg_running checks if monitor is running, attempts to restart if it is not.
+ * The function verifies if the extension version is the same as expected.  It returns true
+ * if the monitor is up and running.
+ */
+bool
+ensure_monitor_pg_running(Monitor *monitor, struct MonitorConfig *mconfig)
+{
+	MonitorExtensionVersion version = { 0 };
+	char postgresUri[MAXCONNINFO];
+
+	if (!pg_is_running(mconfig->pgSetup.pg_ctl, mconfig->pgSetup.pgdata))
+	{
+		log_info("Postgres is not running, starting postgres");
+		pgsql_finish(&(monitor->pgsql));
+
+		if (!pg_ctl_start(mconfig->pgSetup.pg_ctl,
+						  mconfig->pgSetup.pgdata,
+						  mconfig->pgSetup.pgport,
+						  mconfig->pgSetup.listen_addresses))
+		{
+			log_error("Failed to start PostgreSQL, see above for details");
+			return false;
+		}
+
+		/*
+		 * Check version compatibility.
+		 *
+		 * The function terminates any existing connection during clenaup
+		 * therefore it is not called when PG is found to be running to not
+		 * to intervene pgsql_listen call.
+		 */
+		if (!monitor_ensure_extension_version(monitor, &version))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+	}
+
+	log_debug("PostgreSQL is running in \"%s\" on port %d",
+			  mconfig->pgSetup.pgdata, mconfig->pgSetup.pgport);
```

Comment:
I think we're going to display that at every iteration of the LISTEN loop, is that right? we should move that to just before the loop in the `monitor_service_run` function, right?

Reply:
it used to be log_info. I was trying to reduce the frequency. Okay with moving this line outside.

Reply:
maybe just remove it altogether